tests/v2/validation/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,450 @@
+	"github.com/rancher/rancher/tests/framework/pkg/environmentflag"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+	"github.com/rancher/rancher/tests/integration/pkg/defaults"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type UpgradeTestSuite struct {
+	suite.Suite
+	session   *session.Session
+	client    *rancher.Client
+	project   *management.Project
+	namespace *corev1.Namespace
+	names     *resourceNames
+}
+
+func (u *UpgradeTestSuite) TearDownSuite() {
+	u.session.Cleanup()
+}
+
+func (u *UpgradeTestSuite) SetupSuite() {
+	testSession := session.NewSession(u.T())
+	u.session = testSession
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(u.T(), err)
+
+	u.client = client
+
+	names := newNames()
+	u.names = names
+
+	clusterName := client.RancherConfig.ClusterName
+	require.NotEmptyf(u.T(), clusterName, "Cluster name to install resources is not set")
+
+	clusterID, err := clusters.GetClusterIDByName(client, clusterName)
+	require.NoError(u.T(), err)
+
+	projectName := names.core["projectName"]
+	project, err := projects.GetProjectByName(client, clusterID, projectName)
+	require.NoError(u.T(), err)
+
+	u.project = project
+
+	if project == nil {
+		projectConfig := &management.Project{
+			ClusterID: clusterID,
+			Name:      projectName,
+		}
+		createdProject, err := client.Management.Project.Create(projectConfig)
+		require.NoError(u.T(), err)
+		require.Equal(u.T(), createdProject.Name, projectName)
+		u.project = createdProject
+	}
+}
+
+func (u *UpgradeTestSuite) TestPreUpgrade() {
+	subSession := u.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := u.client.WithSession(subSession)
+	require.NoError(u.T(), err)
+
+	u.T().Log("Creating a namespace")
+	createdNamespace, err := namespaces.CreateNamespace(client, u.names.random["namespaceName"], "{}", map[string]string{}, map[string]string{}, u.project)
+	require.NoError(u.T(), err)
+	assert.Equal(u.T(), createdNamespace.Name, u.names.random["namespaceName"])
+	u.namespace = createdNamespace
+
+	podTemplate := corev1.PodTemplateSpec{
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				*testContainer,
+			},
+		},
+	}
+
+	u.T().Log("Creating a deployment with the test container")
+	createdDeployment, err := deployments.CreateDeployment(client, u.project.ClusterID, u.names.random["deploymentName"], u.namespace.Name, podTemplate)
+	require.NoError(u.T(), err)
+	assert.Equal(u.T(), createdDeployment.Name, u.names.random["deploymentName"])
+
+	u.T().Log("Waiting deployments to have expected number of available replicas")
+	err = charts.WatchAndWaitDeployments(client, u.project.ClusterID, u.namespace.Name, metav1.ListOptions{})
+	require.NoError(u.T(), err)
+
+	u.T().Log("Creating a daemonset with the test container")
```

Comment:
Great catch, this is not exactly what you're asking for, but this can be addressed and enhanced with this [function](https://github.com/rancher/rancher/blob/release/v2.6/tests/framework/extensions/charts/charts.go#L202-L204).

We need a function that distinguishes workers to achieve what you're asking for. And on top of that, we would need a helper between that list of workers and daemon sets statuses. @igomez06 would this be the correct way to do it?

Reply:
Yes, if you use the client.Management.Nodes api call to get node info, it should tell you if the node is a worker node or not, and then you can compare the number of workers with daemonset replicas.

Reply:
Updated: Added watch and wait for daemon sets, and as the next step added validation of daemon set replicas are equal to the number of workers in the cluster